lib/mongoid/contextual/mongo.rb

```diff
@@ -1038,6 +1043,16 @@ def process_raw_docs(raw_docs, limit)
         limit ? docs : docs.first
       end
 
+      # Queries whether the current context is valid for use with
+      # the #count_documents? predicate. A context is valid if it
+      # does not include a `$where` operator.
+      #
+      # @return [ true | false ] whether or not the current context
+      #   includes a `$where` operator.
+      def valid_for_count_documents?
+        view.filter.keys.none? { |k| k == '$where' }
+      end
```

Comment:
This works for any `$where` on the top level. However, `$where` may be nested within an `$and`, `$or`, `$not`, or `$nor`. Consider:
```js
{
    "$and": [
        { "foo": { "$eq": 1 } },
        { "$where": function() { return (hex_md5(this.name) == "9b53e667f30cd329dca1ec9e6a83e994") } }
    ]
}
```

I'm not sure whether it's possible to chain `for_js` with other operators to create such a condition, but if so this method may need to traverse any of the logical operators and check if they are each valid.

Reply:
Yeah, I wondered about that. I'll take another look.

@alexbevi -- I love that judo, but I worry about potential false positives (e.g. a string literal in the query containing `"$where"` ). Such false positives might not matter, though; the worst case is that it uses the deprecated `count` API.

Reply:
Yeah, since we're erring on the side of caution with this and most likely aren't affected by the edge cases of `count` providing wrong results, I think we're fine using `count` when the query would also have worked with `countDocuments`.